File: Assigment 2/my-app/Firebase/metricTracker.js
const { Builder } = require('selenium-webdriver');
const fs = require('fs');

async function writeToCSV(filename, metrics) {
    const header = Object.keys(metrics[0]);
    const rows = metrics.map(metric => Object.values(metric));
    const csv = [header, ...rows].map(row => row.join(',')).join('\n');
    await fs.promises.writeFile(filename, csv);
}

async function main() {
    const driver = await new Builder().forBrowser('chrome').build();

    await driver.get("http://localhost:3000/");
    const metrics = [];
    const SAMPLE_SIZE = 10;
    let count = 0;

    const start_time = Date.now() / 1000;
    while (count < SAMPLE_SIZE) {
        let presence_time = start_time;
        const current_time = Date.now() / 1000;
        presence_time = current_time - start_time;
        console.log(`Presence time: ${presence_time} seconds`);

        const scroll_height = await driver.executeScript("return document.body.scrollHeight");
        const current_scroll = await driver.executeScript("return window.pageYOffset");
        console.log(`Scrolled ${current_scroll}/${scroll_height} pixels`);

        metrics.push({
            "TIMESTAMP (HH:MM:SS)": new Date().toLocaleTimeString(),
            "Precence time (Seconds)": presence_time,
            "Scrolling (Pixels)": current_scroll / scroll_height
        });
        count += 1;
        await driver.sleep(2000);
    }
    await driver.quit();
    console.log(metrics);
    await writeToCSV("metrics.csv", metrics);
}

main();

File: Assignment 3 Database/my-app/metric_tracker.py
import time
from selenium import webdriver
import collections
import csv

def writeToCSV(filename : str, metrics : dict):
    with open(file=filename, mode="w", newline="") as fp:
        #create writer object
        writer = csv.DictWriter(fp, fieldnames=metrics[0].keys())

        writer.writeheader()
        for metric in metrics:
            writer.writerow(metric)

def main():
    # Initialize browser
    driver = webdriver.Chrome()
    # Navigate to your website 
    driver.get("http://localhost:3000/")
    metrics = [] #collections.defaultdict(list)

    SAMPLE_SIZE = 10
    count = 0
    # Track presence time 
    start_time = time.time()
    while count < SAMPLE_SIZE:
        presence_time = start_time
        current_time = time.time()
        presence_time = current_time - start_time
        print(f"Presence time: {presence_time} seconds")
        #metrics["Presence time (Seconds)"].append (presence_time)
    
        # TIMESTAMP : "Present Time (Seconds)" : Presence Time
        # TIMESTAMP : "Scrolling (Pixels)" : Scroll
        # Track scrolling
        scroll_height = driver.execute_script("return document.body.scrollHeight")  
        current_scroll = driver.execute_script("return window.pageYOffset")
        print(f"Scrolled {current_scroll}/{scroll_height} pixels")
        #metrics["Scrolling (Pixels)"].append (current_scroll/scroll_height)
        metrics.append({"TIMESTAMP (HH:MM:SS)": time.strftime("%H:%M:%S", time.localtime()),
                        "Precence time (Seconds)" : presence_time,
                        "Scrolling (Pixels)" : current_scroll/scroll_height })
    
        count += 1
        time.sleep(2) 
        
    driver.quit()
    print(metrics)
    writeToCSV("metrics.csv", metrics)

if __name__=="__main__":
    main()